core/src/main/scala/cats/data/Inverse.scala

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2015 Typelevel
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package cats
+package data
+
+import cats.kernel.LowerBounded
+import cats.kernel.UpperBounded
+
+object InverseImpl extends InverseInstances with Newtype {
+  private[data] def unwrap[A](value: Type[A]): A =
+    value.asInstanceOf[A]
+
+  private[data] def create[A](value: A): Type[A] =
+    value.asInstanceOf[Type[A]]
+
+  def apply[A](value: A): Inverse[A] =
+    create(value)
+
+  implicit def catsInverseOps[A](value: Inverse[A]): InverseOps[A] =
+    new InverseOps[A](value)
+}
+
+sealed class InverseOps[A](val inverse: Inverse[A]) {
+  def value: A = InverseImpl.unwrap(inverse)
+}
+
+sealed abstract private[data] class InverseInstances extends InverseInstances0 {
+  implicit def hashAndOrderForInverse[A: Hash: Order]: Hash[Inverse[A]] with Order[Inverse[A]] =
+    new Hash[Inverse[A]] with Order[Inverse[A]] {
+      override def hash(x: Inverse[A]): Int =
+        Hash[A].hash(x.value)
```

Comment:
Looks like `Hash[Inverse[A]].hash <==> Hash[A].hash`
so what is the point of having an `Inverse` for `Hash` then?

Reply:
There isn't if you know you're dealing with an `Inverse`. This is provided to solve constraints in a generic context, e.g. `def foo[A: Hash]` I want to be able to call that with `A` and `Inverse[A]` for all `A` which have a `Hash`.